tests/ZichtTest/Bundle/FrameworkExtraBundle/JsonSchema/SchemaServiceTest.php

```diff
@@ -0,0 +1,143 @@
+
+use Swaggest\JsonSchema\Schema;
+use Zicht\Bundle\FrameworkExtraBundle\JsonSchema\SchemaService;
+
+class SchemaServiceTest extends \PHPUnit_Framework_TestCase
+{
+    /** @var SchemaService */
+    protected $schemaService = null;
+
+    protected function setUp()
+    {
+        $this->schemaService = new SchemaService(__DIR__);
+    }
+
+    /**
+     * @param Schema|string|object|array $schema
+     * @param null|string|int|object|array $data
+     * @param null|string|int|object|array $expected
+     * @param null|string $expectedMessageRegExp
+     * @dataProvider validateSchemaTypesProvider
+     * @dataProvider validateProvider
+     */
+    public function testValidate($schema, $data, $expected, $expectedMessageRegExp = null)
+    {
+        $result = $this->schemaService->validate($schema, $data, $message);
+        if ($expectedMessageRegExp === null) {
+            self::assertNull($message);
+        } else {
+            self::assertRegExp($expectedMessageRegExp, $message);
+        }
+        self::assertEquals($expected, $result);
+    }
+
+    /**
+     * @param Schema|string|object|array $schema
+     * @param null|string|int|object|array $data
+     * @param null|string|int|object|array $expected
+     * @param null|string $expectedMessageRegExp
+     * @dataProvider migrateProvider
+     */
+    public function testMigrate($schema, $data, $expected, $expectedMessageRegExp = null)
+    {
+        $result = $this->schemaService->migrate($schema, $data, $message);
+        if ($expectedMessageRegExp === null) {
+            self::assertNull($message);
+        } else {
+            self::assertRegExp($expectedMessageRegExp, $message);
+        }
+        self::assertEquals($expected, $result);
+    }
+
+    /**
+     * Provides tests to ensure that the $schema provided to SchemaService::validate can be Schema, string, or object
+     *
+     * @return array[]
+     * @throws \Swaggest\JsonSchema\Exception
+     * @throws \Swaggest\JsonSchema\InvalidValue
+     */
+    public function validateSchemaTypesProvider()
+    {
+        return [
+            //
+            // Schema can be a Schema instance
+            //
+            [Schema::import(__DIR__ . '/bundles/validate-foo.schema.json'), 'foo', true],
+
+            //
+            // Schema can be a string
+            //
+            // Allows relative path without starting '/'
+            ['bundles/validate-foo.schema.json', 'foo', true],
+            // Allows relative path with starting '/'
+            ['/bundles/validate-foo.schema.json', 'foo', true],
+            // Allows absolute path
+            [__DIR__ . '/bundles/validate-foo.schema.json', 'foo', true],
+
+            //
+            // Schema can be an object
+            //
+            [json_decode(file_get_contents(__DIR__ . '/bundles/validate-foo.schema.json')), 'foo', true],
+        ];
+    }
+
+    /**
+     * Provides tests that check that a given value validates with the given schema
```

Comment:
Is dat niet de verantwoordelijkheid van de Json-schema library? Het enige dat wij zelf toevoegen in de `validate` methode is een datatransformatie, een regel met encoding en een try/catch, moet je niet daarop focusen om de werking daarvan aan te tonen?

Reply:
We doen wat dingen om de `[]` vs `{}` te fixen, iets waar php problemen mee heeft.  Dus het is wel nuttig om wat eenvoudige testjes hiervoor te hebben